### PR TITLE
added Scons modules #define generator

### PR DESCRIPTION
**Pull request by jeronimo-schreyer**

### CODE REVIEW

**Comment by jeronimo-schreyer** [comment]
Generated file describes what modules are available

**Comment by akien-mga** [comment]
There's already code to define `MODULE_*_ENABLED` right next to where you generate this header, the only difference is that it's passed only to `env_modules`, and not to the main `env`. So modules can know which other modules are enabled, but `scene/` or `editor/` code doesn't.

The main reason for that is that passing all `MODULE_*_ENABLED` defines to the command line would likely double the length of an already quite long compiler call. Generate a header with those defines can indeed be an alternative, but if we go this route, there's no reason to keep passing the defines to `env_modules`.

There's also `SVG_ENABLED` and `FREETYPE_ENABLED` defines passed to the main `env` in their respective module, which should also be ported to `MODULE_*_ENABLED` via a generated header if we go this route.

All in all this is more something for thorough proposal on how to refactor the way module dependencies are handled. Your PR lacks description so it's not clear what you want to achieve, but if the aim is just to allow disabling regex support, a simpler fix would be to define append `REGEX_ENABLED` to the env as done for SVG and FREETYPE, and we can further discuss how to refactor all this for 4.0.

File: scene/gui/rich_text_label.cpp [hunk 1]
```diff
@@ -29,6 +29,7 @@
 /*************************************************************************/
 
 #include "rich_text_label.h"
+#include "../../modules/modules_enabled.gen.h"
```
Comment (akien-mga) [inline scene/gui/rich_text_label.cpp]: This should be below together with other includes. It should also start from the root, we don't use relative paths.
`modules/regex/regex.h` should not be included if the module is disabled.

**Comment by Xrayez** [comment]
> The main reason for that is that passing all MODULE_*_ENABLED defines to the command line would likely double the length of an already quite long compiler call. 

I wanted to do this while attempting to not compile Bullet-related code if the module is disabled. In fact I've successfully done this, except I realized that there was quite a bunch of places already to wrap with MODULE_BULLET_ENABLED. The idea has originated from #32216.

> There's also SVG_ENABLED and FREETYPE_ENABLED defines passed to the main env in their respective module, which should also be ported to MODULE_*_ENABLED via a generated header if we go this route.

Also GDSCRIPT_ENABLED.

I think this needs [godot-proposal](https://github.com/godotengine/godot-proposals/) again. 🤔

**Comment by punto-** [comment]
I like this as an alternative to passing MODULE_*_ENABLED (but just to be clear @jeronimo-schreyer works for me). I never really liked adding those defines outside of modules/ and in reality almost none of the modules use them, except that ones that use other modules, like ogg and vorbis, and it makes sense for code outside modules/ to sometimes need to know if a module is there. This header would allow anyone to know about the modules, without having to pollute the compile command (and causing a full rebuild when one module is enabled/disabled). This doesn't remove the defines from the command line to keep compatibility, it's just to address the issue of regexp used in the RichTextLabel (regexp can't be used in consoles because it has a jit)

File: modules/SCsub [hunk 2]
```diff
@@ -10,12 +10,22 @@ env.modules_sources = []
 
 env_modules.add_source_files(env.modules_sources, "register_module_types.gen.cpp")
 
+def generate_modules_enabled(target = None, source = None, env = None):
+	with open(target[0].path, 'w') as f:
```
Comment (akien-mga) [inline modules/SCsub]: Should use 4 spaces for indentation.

**Comment by akien-mga** [comment]
> This doesn't remove the defines from the command line to keep compatibility, it's just to address the issue of regexp used in the RichTextLabel (regexp can't be used in consoles because it has a jit)

IMO this change is fine, but then the old style should be removed, we don't need to keep both. It only breaks compatibility for thirdparty modules that would happen to reuse our own `env_modules`, as it's the only one which is aware of `MODULE_*_ENABLED` macros so far, and then only those modules who actively use features provided by modules *and* thought about adding checks. So all in all it's probably a relatively safe change.

**Comment by akien-mga** [comment]
There's one potential issue with this approach though, which is that any time a module is enabled or disabled, the header will change and this will force recompilation of all the files that include it. This could be particularly problematic if we also remove the hardcoded `MINIZIP_ENABLED` and include `modules/modules_enabled.gen.h` in `core/`.

So we may need to tread cautiously here as our incremental builds are already a bit on the edge.

**Comment by jeronimo-schreyer** [comment]
I squashed previous commits into one

**Comment by akien-mga** [comment]
Superseded by #35963. Thanks for the contribution nevertheless, which was used as basis for that bigger PR.

**Pull request closed without merging**